File: src/operations/control.py
"""
Control flow operations handler for the auto-workflow system.
"""
import re
from typing import Dict, Any, Union
from .handler import Handler
from ..models import ExecutionContext


class ControlHandler(Handler):
    """Handles control flow operations."""

    def get_type(self) -> str:
        """Return the operation type."""
        return "control"

    def validate(self, params: Dict[str, Any]) -> None:
        """Check if the parameters are valid."""
        action = params.get("action")
        if not action or not isinstance(action, str):
            raise ValueError("action parameter is required")

        if action == "set_variable":
            if "name" not in params:
                raise ValueError("name parameter is required for set_variable action")
            if "value" not in params:
                raise ValueError("value parameter is required for set_variable action")
        elif action == "get_variable":
            if "name" not in params:
                raise ValueError("name parameter is required for get_variable action")
        elif action == "increment":
            if "name" not in params:
                raise ValueError("name parameter is required for increment action")
        elif action == "decrement":
            if "name" not in params:
                raise ValueError("name parameter is required for decrement action")
        elif action == "append":
            if "name" not in params:
                raise ValueError("name parameter is required for append action")
            if "value" not in params:
                raise ValueError("value parameter is required for append action")
        elif action == "log":
            if "message" not in params:
                raise ValueError("message parameter is required for log action")
        elif action == "assert":
            if "condition" not in params:
                raise ValueError("condition parameter is required for assert action")

    def execute(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Perform the control operation."""
        action = params["action"]

        if action == "set_variable":
            return self._execute_set_variable(params, exec_ctx)
        elif action == "get_variable":
            return self._execute_get_variable(params, exec_ctx)
        elif action == "increment":
            return self._execute_increment(params, exec_ctx)
        elif action == "decrement":
            return self._execute_decrement(params, exec_ctx)
        elif action == "append":
            return self._execute_append(params, exec_ctx)
        elif action == "log":
            return self._execute_log(params, exec_ctx)
        elif action == "assert":
            return self._execute_assert(params, exec_ctx)
        else:
            raise ValueError(f"unsupported control action: {action}")

    def _execute_set_variable(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Execute set_variable operation."""
        name = self._get_string_param(params, "name")
        value = params["value"]

        # Resolve variable references if needed
        resolved_value = self._resolve_value(value, exec_ctx)

        if exec_ctx.variables is None:
            exec_ctx.variables = {}

        exec_ctx.variables[name] = resolved_value

        return {
            "success": True,
            "name": name,
            "value": resolved_value,
        }

    def _execute_get_variable(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Execute get_variable operation."""
        name = self._get_string_param(params, "name")

        if name not in exec_ctx.variables:
            raise ValueError(f"variable '{name}' not found")

        value = exec_ctx.variables[name]

        return {
            "success": True,
            "name": name,
            "value": value,
        }

    def _execute_increment(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Execute increment operation."""
        name = self._get_string_param(params, "name")
        amount = self._get_int_param(params, "amount", 1)

        if name not in exec_ctx.variables:
            raise ValueError(f"variable '{name}' not found")

        value = exec_ctx.variables[name]

        if isinstance(value, (int, float)):
            new_value = value + amount
        elif isinstance(value, str):
            try:
                new_value = float(value) + amount
            except ValueError:
                raise ValueError(f"variable '{name}' is not a number")
        else:
            raise ValueError(f"variable '{name}' is not a number")

        exec_ctx.variables[name] = new_value

        return {
            "success": True,
            "name": name,
            "value": new_value,
        }

    def _execute_decrement(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Execute decrement operation."""
        name = self._get_string_param(params, "name")
        amount = self._get_int_param(params, "amount", 1)

        if name not in exec_ctx.variables:
            raise ValueError(f"variable '{name}' not found")

        value = exec_ctx.variables[name]

        if isinstance(value, (int, float)):
            new_value = value - amount
        elif isinstance(value, str):
            try:
                new_value = float(value) - amount
            except ValueError:
                raise ValueError(f"variable '{name}' is not a number")
        else:
            raise ValueError(f"variable '{name}' is not a number")

        exec_ctx.variables[name] = new_value

        return {
            "success": True,
            "name": name,
            "value": new_value,
        }

    def _execute_append(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Execute append operation."""
        name = self._get_string_param(params, "name")
        value = params["value"]

        # Resolve variable references if needed
        resolved_value = self._resolve_value(value, exec_ctx)

        if name not in exec_ctx.variables:
            raise ValueError(f"variable '{name}' not found")

        existing_value = exec_ctx.variables[name]

        if isinstance(existing_value, list):
            exec_ctx.variables[name] = existing_value + [resolved_value]
        else:
            raise ValueError(f"variable '{name}' is not an array")

        return {
            "success": True,
            "name": name,
            "appended": resolved_value,
        }

    def _execute_log(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Execute log operation."""
        message = self._get_string_param(params, "message")
        level = self._get_string_param(params, "level", "info")

        # Resolve variable references in message
        resolved_message = self._resolve_variables(message, exec_ctx)

        print(f"[{level.upper()}] {resolved_message}")

        return {
            "success": True,
            "level": level,
            "message": resolved_message,
        }

    def _execute_assert(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Execute assert operation."""
        condition = params["condition"]

        # Evaluate the condition
        result = self._evaluate_condition(condition, exec_ctx)

        if not result:
            message = self._get_string_param(params, "message", "assertion failed")
            raise AssertionError(f"assertion failed: {message}")

        return {
            "success": True,
            "condition": condition,
        }

    def _resolve_value(self, value: Any, exec_ctx: ExecutionContext) -> Any:
        """Resolve variable references in a value."""
        if isinstance(value, str):
            # Check for variable reference like ${variable_name}
            match = re.match(r'\$\{(.+?)\}', value)
            if match:
                var_name = match.group(1)
                if var_name in exec_ctx.variables:
                    return exec_ctx.variables[var_name]
                else:
                    raise ValueError(f"variable '{var_name}' not found")
            return value
        elif isinstance(value, dict):
            result = {}
            for k, v in value.items():
                result[k] = self._resolve_value(v, exec_ctx)
            return result
        elif isinstance(value, list):
            return [self._resolve_value(item, exec_ctx) for item in value]
        else:
            return value

    def _resolve_variables(self, text: str, exec_ctx: ExecutionContext) -> str:
        """Resolve variable references in text."""
        result = text

        # Find all variable references
        while True:
            match = re.search(r'\$\{(.+?)\}', result)
            if not match:
                break

            var_name = match.group(1)
            if var_name in exec_ctx.variables:
                var_value = str(exec_ctx.variables[var_name])
                result = result[:match.start()] + var_value + result[match.end():]
            else:
                # Keep the reference if variable not found
                result = result[:match.end()] + result[match.end():]

        return result

    def _evaluate_condition(self, condition: Any, exec_ctx: ExecutionContext) -> bool:
        """Evaluate a condition."""
        if isinstance(condition, bool):
            return condition
        elif isinstance(condition, str):
            # Try to resolve variable reference
            match = re.match(r'\$\{(.+?)\}', condition)
            if match:
                var_name = match.group(1)
                if var_name in exec_ctx.variables:
                    value = exec_ctx.variables[var_name]
                    # For non-boolean values, check if truthy
                    if isinstance(value, bool):
                        return value
                    elif isinstance(value, (int, float)):
                        return value != 0
                    elif isinstance(value, str):
                        return value != ""
                    else:
                        return value is not None
                else:
                    return False
            # Try to parse as boolean
            if condition.lower() == "true":
                return True
            elif condition.lower() == "false":
                return False
            else:
                raise ValueError(f"invalid condition: {condition}")
        elif isinstance(condition, dict):
            # Complex condition with operator
            return self._evaluate_complex_condition(condition, exec_ctx)
        elif isinstance(condition, (int, float)):
            # Numeric values: true if non-zero
            return condition != 0
        else:
            raise ValueError(f"invalid condition type: {type(condition)}")

    def _evaluate_complex_condition(self, condition: Dict[str, Any], exec_ctx: ExecutionContext) -> bool:
        """Evaluate a complex condition with operator."""
        operator = condition.get("operator")
        if not operator:
            raise ValueError("operator is required for complex condition")

        left = self._resolve_value(condition.get("left"), exec_ctx)
        right = self._resolve_value(condition.get("right"), exec_ctx)

        if operator == "equals":
            return self._compare_values(left, right) == 0
        elif operator == "not_equals":
            return self._compare_values(left, right) != 0
        elif operator == "greater":
            return self._compare_values(left, right) > 0
        elif operator == "less":
            return self._compare_values(left, right) < 0
        elif operator == "greater_equal":
            return self._compare_values(left, right) >= 0
        elif operator == "less_equal":
            return self._compare_values(left, right) <= 0
        elif operator == "contains":
            return self._contains_value(left, right)
        elif operator == "and":
            left_result = self._evaluate_condition(left, exec_ctx)
            if not left_result:
                return False
            return self._evaluate_condition(right, exec_ctx)
        elif operator == "or":
            left_result = self._evaluate_condition(left, exec_ctx)
            if left_result:
                return True
            return self._evaluate_condition(right, exec_ctx)
        elif operator == "not":
            return not self._evaluate_condition(left, exec_ctx)
        else:
            raise ValueError(f"unsupported operator: {operator}")

    def _compare_values(self, a: Any, b: Any) -> int:
        """Compare two values. Returns -1, 0, or 1."""
        # Try to convert to numbers
        a_num = self._to_float(a)
        b_num = self._to_float(b)

        if a_num is not None and b_num is not None:
            if a_num < b_num:
                return -1
            elif a_num > b_num:
                return 1
            else:
                return 0

        # Fall back to string comparison
        a_str = str(a)
        b_str = str(b)

        if a_str < b_str:
            return -1
        elif a_str > b_str:
            return 1
        else:
            return 0

    def _to_float(self, value: Any) -> Union[float, None]:
        """Convert value to float if possible."""
        if isinstance(value, (int, float)):
            return float(value)
        elif isinstance(value, str):
            try:
                return float(value)
            except ValueError:
                return None
        else:
            return None

    def _contains_value(self, container: Any, value: Any) -> bool:
        """Check if container contains value."""
        if isinstance(container, str):
            return str(value) in container
        elif isinstance(container, list):
            return value in container
        else:
            return False

    # Helper methods
    def _get_string_param(self, params: Dict[str, Any], key: str, default: str = "") -> str:
        """Get string parameter."""
        value = params.get(key)
        if isinstance(value, str):
            return value
        return str(value) if value is not None else default

    def _get_int_param(self, params: Dict[str, Any], key: str, default: int = 0) -> int:
        """Get integer parameter."""
        value = params.get(key)
        if value is None:
            return default
        if isinstance(value, int):
            return value
        if isinstance(value, float):
            return int(value)
        if isinstance(value, str):
            try:
                return int(value)
            except ValueError:
                return default
        return default

File: pkg/operations/control.go
package operations

import (
	"auto-workflow/pkg/models"
	"context"
	"fmt"
	"reflect"
	"strconv"
	"strings"
)

// ControlHandler handles control flow operations
type ControlHandler struct{}

// NewControlHandler creates a new control handler
func NewControlHandler() *ControlHandler {
	return &ControlHandler{}
}

// GetType returns the operation type
func (h *ControlHandler) GetType() string {
	return "control"
}

// Validate checks if the parameters are valid
func (h *ControlHandler) Validate(params map[string]interface{}) error {
	action, ok := params["action"].(string)
	if !ok {
		return fmt.Errorf("action parameter is required")
	}

	switch action {
	case "set_variable":
		if _, ok := params["name"]; !ok {
			return fmt.Errorf("name parameter is required for set_variable action")
		}
		if _, ok := params["value"]; !ok {
			return fmt.Errorf("value parameter is required for set_variable action")
		}
	case "get_variable":
		if _, ok := params["name"]; !ok {
			return fmt.Errorf("name parameter is required for get_variable action")
		}
	case "increment":
		if _, ok := params["name"]; !ok {
			return fmt.Errorf("name parameter is required for increment action")
		}
	case "decrement":
		if _, ok := params["name"]; !ok {
			return fmt.Errorf("name parameter is required for decrement action")
		}
	case "append":
		if _, ok := params["name"]; !ok {
			return fmt.Errorf("name parameter is required for append action")
		}
		if _, ok := params["value"]; !ok {
			return fmt.Errorf("value parameter is required for append action")
		}
	case "log":
		if _, ok := params["message"]; !ok {
			return fmt.Errorf("message parameter is required for log action")
		}
	case "assert":
		if _, ok := params["condition"]; !ok {
			return fmt.Errorf("condition parameter is required for assert action")
		}
	}

	return nil
}

// Execute performs the control operation
func (h *ControlHandler) Execute(ctx context.Context, params map[string]interface{}, execCtx interface{}) (map[string]interface{}, error) {
	action := params["action"].(string)

	// Get execution context
	context, ok := execCtx.(*models.ExecutionContext)
	if !ok {
		return nil, fmt.Errorf("invalid execution context")
	}

	var result map[string]interface{}
	var err error

	switch action {
	case "set_variable":
		result, err = h.executeSetVariable(ctx, params, context)
	case "get_variable":
		result, err = h.executeGetVariable(ctx, params, context)
	case "increment":
		result, err = h.executeIncrement(ctx, params, context)
	case "decrement":
		result, err = h.executeDecrement(ctx, params, context)
	case "append":
		result, err = h.executeAppend(ctx, params, context)
	case "log":
		result, err = h.executeLog(ctx, params, context)
	case "assert":
		result, err = h.executeAssert(ctx, params, context)
	default:
		return nil, fmt.Errorf("unsupported control action: %s", action)
	}

	return result, err
}

func (h *ControlHandler) executeSetVariable(ctx context.Context, params map[string]interface{}, execCtx *models.ExecutionContext) (map[string]interface{}, error) {
	name := getStringParam(params, "name")
	value := params["value"]

	// Resolve variable references if needed
	resolvedValue, err := h.resolveValue(value, execCtx)
	if err != nil {
		return nil, fmt.Errorf("resolve value failed: %v", err)
	}

	if execCtx.Variables == nil {
		execCtx.Variables = make(map[string]interface{})
	}

	execCtx.Variables[name] = resolvedValue

	return map[string]interface{}{
		"success": true,
		"name":    name,
		"value":   resolvedValue,
	}, nil
}

func (h *ControlHandler) executeGetVariable(ctx context.Context, params map[string]interface{}, execCtx *models.ExecutionContext) (map[string]interface{}, error) {
	name := getStringParam(params, "name")

	value, exists := execCtx.Variables[name]
	if !exists {
		return nil, fmt.Errorf("variable '%s' not found", name)
	}

	return map[string]interface{}{
		"success": true,
		"name":    name,
		"value":   value,
	}, nil
}

func (h *ControlHandler) executeIncrement(ctx context.Context, params map[string]interface{}, execCtx *models.ExecutionContext) (map[string]interface{}, error) {
	name := getStringParam(params, "name")
	amount := getIntParamDefault(params, "amount", 1)

	value, exists := execCtx.Variables[name]
	if !exists {
		return nil, fmt.Errorf("variable '%s' not found", name)
	}

	var newValue int
	switch v := value.(type) {
	case float64:
		newValue = int(v) + amount
	case int:
		newValue = v + amount
	default:
		return nil, fmt.Errorf("variable '%s' is not a number", name)
	}

	execCtx.Variables[name] = newValue

	return map[string]interface{}{
		"success": true,
		"name":    name,
		"value":   newValue,
	}, nil
}

func (h *ControlHandler) executeDecrement(ctx context.Context, params map[string]interface{}, execCtx *models.ExecutionContext) (map[string]interface{}, error) {
	name := getStringParam(params, "name")
	amount := getIntParamDefault(params, "amount", 1)

	value, exists := execCtx.Variables[name]
	if !exists {
		return nil, fmt.Errorf("variable '%s' not found", name)
	}

	var newValue int
	switch v := value.(type) {
	case float64:
		newValue = int(v) - amount
	case int:
		newValue = v - amount
	default:
		return nil, fmt.Errorf("variable '%s' is not a number", name)
	}

	execCtx.Variables[name] = newValue

	return map[string]interface{}{
		"success": true,
		"name":    name,
		"value":   newValue,
	}, nil
}

func (h *ControlHandler) executeAppend(ctx context.Context, params map[string]interface{}, execCtx *models.ExecutionContext) (map[string]interface{}, error) {
	name := getStringParam(params, "name")
	value := params["value"]

	// Resolve variable references if needed
	resolvedValue, err := h.resolveValue(value, execCtx)
	if err != nil {
		return nil, fmt.Errorf("resolve value failed: %v", err)
	}

	existingValue, exists := execCtx.Variables[name]
	if !exists {
		return nil, fmt.Errorf("variable '%s' not found", name)
	}

	switch arr := existingValue.(type) {
	case []interface{}:
		execCtx.Variables[name] = append(arr, resolvedValue)
	default:
		return nil, fmt.Errorf("variable '%s' is not an array", name)
	}

	return map[string]interface{}{
		"success":  true,
		"name":     name,
		"appended": resolvedValue,
	}, nil
}

func (h *ControlHandler) executeLog(ctx context.Context, params map[string]interface{}, execCtx *models.ExecutionContext) (map[string]interface{}, error) {
	message := getStringParam(params, "message")
	level := getStringParamDefault(params, "level", "info")

	// Resolve variable references in message
	resolvedMessage, err := h.resolveVariables(message, execCtx)
	if err != nil {
		return nil, fmt.Errorf("resolve variables failed: %v", err)
	}

	fmt.Printf("[%s] %s\n", strings.ToUpper(level), resolvedMessage)

	return map[string]interface{}{
		"success": true,
		"level":   level,
		"message": resolvedMessage,
	}, nil
}

func (h *ControlHandler) executeAssert(ctx context.Context, params map[string]interface{}, execCtx *models.ExecutionContext) (map[string]interface{}, error) {
	condition := params["condition"]

	// Evaluate the condition
	result, err := h.evaluateCondition(condition, execCtx)
	if err != nil {
		return nil, fmt.Errorf("evaluate condition failed: %v", err)
	}

	if !result {
		message := getStringParamDefault(params, "message", "assertion failed")
		return nil, fmt.Errorf("assertion failed: %s", message)
	}

	return map[string]interface{}{
		"success":   true,
		"condition": condition,
	}, nil
}

// Helper functions

func (h *ControlHandler) resolveValue(value interface{}, execCtx *models.ExecutionContext) (interface{}, error) {
	switch v := value.(type) {
	case string:
		if strings.HasPrefix(v, "${") && strings.HasSuffix(v, "}") {
			varName := v[2 : len(v)-1]
			if val, exists := execCtx.Variables[varName]; exists {
				return val, nil
			}
			return nil, fmt.Errorf("variable '%s' not found", varName)
		}
		return v, nil
	case map[string]interface{}:
		result := make(map[string]interface{})
		for k, val := range v {
			resolved, err := h.resolveValue(val, execCtx)
			if err != nil {
				return nil, err
			}
			result[k] = resolved
		}
		return result, nil
	case []interface{}:
		result := make([]interface{}, len(v))
		for i, val := range v {
			resolved, err := h.resolveValue(val, execCtx)
			if err != nil {
				return nil, err
			}
			result[i] = resolved
		}
		return result, nil
	default:
		return v, nil
	}
}

func (h *ControlHandler) resolveVariables(text string, execCtx *models.ExecutionContext) (string, error) {
	result := text

	// Find all variable references
	start := 0
	for {
		startIdx := strings.Index(result[start:], "${")
		if startIdx == -1 {
			break
		}
		startIdx += start

		endIdx := strings.Index(result[startIdx:], "}")
		if endIdx == -1 {
			break
		}
		endIdx += startIdx

		varName := result[startIdx+2 : endIdx]
		if val, exists := execCtx.Variables[varName]; exists {
			varValue := fmt.Sprintf("%v", val)
			result = result[:startIdx] + varValue + result[endIdx+1:]
			start = startIdx + len(varValue)
		} else {
			start = endIdx + 1
		}
	}

	return result, nil
}

func (h *ControlHandler) evaluateCondition(condition interface{}, execCtx *models.ExecutionContext) (bool, error) {
	switch cond := condition.(type) {
	case bool:
		return cond, nil
	case string:
		// Try to resolve variable reference
		if strings.HasPrefix(cond, "${") && strings.HasSuffix(cond, "}") {
			varName := cond[2 : len(cond)-1]
			if val, exists := execCtx.Variables[varName]; exists {
				// For non-boolean values, return true if the value exists
				// The actual comparison should be done in complex conditions
				if b, ok := val.(bool); ok {
					return b, nil
				}
				// For numeric values, check if it's truthy (non-zero)
				if num, ok := toFloat64(val); ok {
					return num != 0, nil
				}
				// For strings, check if it's not empty
				if str, ok := val.(string); ok {
					return str != "", nil
				}
				// For other types, return true if it exists
				return true, nil
			}
			return false, fmt.Errorf("variable '%s' not found", varName)
		}
		// Try to parse as boolean
		switch strings.ToLower(cond) {
		case "true":
			return true, nil
		case "false":
			return false, nil
		}
		return false, fmt.Errorf("invalid condition: %s", cond)
	case map[string]interface{}:
		// Complex condition with operator
		return h.evaluateComplexCondition(cond, execCtx)
	case float64, int, int32, int64:
		// Numeric values: true if non-zero
		return cond != 0, nil
	default:
		return false, fmt.Errorf("invalid condition type: %v", reflect.TypeOf(condition))
	}
}

func (h *ControlHandler) evaluateComplexCondition(condition map[string]interface{}, execCtx *models.ExecutionContext) (bool, error) {
	operator, ok := condition["operator"].(string)
	if !ok {
		return false, fmt.Errorf("operator is required for complex condition")
	}

	left, err := h.resolveValue(condition["left"], execCtx)
	if err != nil {
		return false, err
	}

	right, err := h.resolveValue(condition["right"], execCtx)
	if err != nil {
		return false, err
	}

	switch operator {
	case "equals":
		// For numeric values, compare after type conversion
		if leftNum, leftOk := toFloat64(left); leftOk {
			if rightNum, rightOk := toFloat64(right); rightOk {
				return leftNum == rightNum, nil
			}
		}
		return reflect.DeepEqual(left, right), nil
	case "not_equals":
		// For numeric values, compare after type conversion
		if leftNum, leftOk := toFloat64(left); leftOk {
			if rightNum, rightOk := toFloat64(right); rightOk {
				return leftNum != rightNum, nil
			}
		}
		return !reflect.DeepEqual(left, right), nil
	case "greater":
		return compareNumbers(left, right) > 0, nil
	case "less":
		return compareNumbers(left, right) < 0, nil
	case "greater_equal":
		return compareNumbers(left, right) >= 0, nil
	case "less_equal":
		return compareNumbers(left, right) <= 0, nil
	case "contains":
		return containsValue(left, right), nil
	case "and":
		leftBool, err := h.evaluateCondition(left, execCtx)
		if err != nil || !leftBool {
			return false, err
		}
		return h.evaluateCondition(right, execCtx)
	case "or":
		leftBool, err := h.evaluateCondition(left, execCtx)
		if err == nil && leftBool {
			return true, nil
		}
		return h.evaluateCondition(right, execCtx)
	case "not":
		result, err := h.evaluateCondition(left, execCtx)
		return !result, err
	default:
		return false, fmt.Errorf("unsupported operator: %s", operator)
	}
}

func compareNumbers(a, b interface{}) int {
	aVal, aOk := toFloat64(a)
	bVal, bOk := toFloat64(b)

	if !aOk || !bOk {
		return 0
	}

	if aVal < bVal {
		return -1
	} else if aVal > bVal {
		return 1
	}
	return 0
}

func toFloat64(v interface{}) (float64, bool) {
	switch val := v.(type) {
	case float64:
		return val, true
	case float32:
		return float64(val), true
	case int:
		return float64(val), true
	case int64:
		return float64(val), true
	case int32:
		return float64(val), true
	case string:
		if f, err := strconv.ParseFloat(val, 64); err == nil {
			return f, true
		}
	}
	return 0, false
}

func containsValue(container, value interface{}) bool {
	switch c := container.(type) {
	case string:
		return strings.Contains(c, fmt.Sprintf("%v", value))
	case []interface{}:
		for _, item := range c {
			if reflect.DeepEqual(item, value) {
				return true
			}
		}
		return false
	default:
		return false
	}
}

// Helper functions for parameter extraction
func getStringParamDefault(params map[string]interface{}, key string, defaultValue string) string {
	if val, ok := params[key]; ok {
		if str, ok := val.(string); ok {
			return str
		}
	}
	return defaultValue
}
